### ISSUE TITLE
Allow covariant type parameters to be implicitly bound to literals

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#5706]
**Is your feature request related to a problem? Please describe.**
Currently literals are erased to their fallback type when used as type parameters.
```py
from typing import Final, reveal_type

class Foo[XT: int]:
    def __init__(self, x: XT):
        self.x: Final = x

reveal_type(Foo(1234567890))  # type is Foo[int]
```

**Describe the solution you’d like**
Ideally, because `XT@Foo` is covariant the type should be `Foo[Literal[1234567890]]`

When using a subclass of `int`, that is reflected in the type
```py
class MyInt(int):
    ...

reveal_type(Foo(MyInt(1234567890)))  # type is Foo[MyInt]
```

In practice, this would help where I have a few classes which are generic over enums which encode the type of the class, only on specific types can operations be performed.
```py
class UserNews[TypeT: UserNewsType = UserNewsType]:
    def __init__(self, news: user_news.GetUserNewsResponse.Event) -> None:
        self.type: Final = cast(TypeT, UserNewsType.try_value(news.eventtype))
        """The type of news entry."""

        self.event_id = news.clan_eventid
        """The event ID this news entry is for. Only used for :attr:`UserNewsType.ScheduledEvent`."""
        self.announcement_id = news.clan_announcementid
        """The announcement ID this news entry is for. Only used for :attr:`UserNewsType.PostedAnnouncement`."""

   def event(self: UserNews[Literal[UserNewsType.ScheduledEvent]]) -> Event:
        return fetch_event(self.event_id)

   def announcement(self: UserNews[Literal[UserNewsType.PostedAnnouncement]]) -> Announcement:
        return fetch_announcement(self.announcement_id)

news_entry: UserNews[Literal[UserNewsType.ScheduledEvent]] = Any
news_entry.event()
news_entry.announcement()  # type error
```

There might be a couple of edge cases using the above example (I'm not sure how pyright would handle them by default without any code changes):
- narrowing is often performed with is for literal enums e.g. `if news_entry.type is UserNewsType.ScheduledEvent: do_something_with_narrowed(news_entry)` but for `int`s you're supposed to use `==`
- match cases would need to support narrowing on the the type parameter

**Comment by erictraut** [comment]
The current behavior is very intentional. Literals are not preserved by the constraint solver unless there's some constraint that requires it. Changing this would break pretty much every code base that currently uses pyright for type checking, and it would result in some very unexpected behaviors. 

```python
class Bar:
    x = Foo(1)

reveal_type(x) # Foo[Literal[1]] ???

Bar.x = Foo(2) # Type error?
```

See [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=constraint-solver-literals).

If you want literals to be retained, you need to introduce a specific type annotation that constrains the type to a literal.

```python
x: Foo[Literal[1]] = Foo(1)
```

**Comment by Gobot1234** [comment]
Would it be possible to add narrowing for them at least?

**Comment by erictraut** [comment]
Pyright already performs narrowing for covariant type arguments. 

```python
x: Foo[Literal[1]] = Foo(1)

y: Foo[int] = x

reveal_type(y) # Foo[Literal[1]]
```

Or did you have something else in mind?

**Comment by Gobot1234** [comment]
Stuff like
```py
if news_entry.type is UserNewsType.ScheduledEvent:
	do_something_with_narrowed(news_entry)
```
or using a match statement

**Comment by erictraut** [comment]
No, the `x is E` type guard doesn't work that that way. It can't narrow the type of an object when the `LHS` of the `is` operator references the attribute of an object because that type variable may be used in other ways (e.g. for the types of other attributes within the class). The type guard can narrow the type of the expression `news_entry.type`, but it can't safely narrow the type of `news_entry`.

**Comment by Gobot1234** [comment]
If the `.type` is bound though by `TypeT` surely the attributes should all have the same type though so it is safe? Do you have an example perhaps where this isn't safe?

**Comment by erictraut** [comment]
```python
from typing import Final


class Foo[XT: int]:
    def __init__(self, x: XT, y: XT):
        self.x: Final = x
        self.y: Final = y


f = Foo(1, 2)
reveal_type(f) # Foo[int]

if f.x is 1:
    reveal_type(f.y) # Not Literal[1]!
```
